course/test/src/matrix3_TEST.cc

```diff
@@ -0,0 +1,173 @@
+    EXPECT_EQ(Matrix3::kIdentity, std::initializer_list<double>({1., 0., 0.,
+        0., 1., 0., 0., 0., 1.}));
+    EXPECT_EQ(Matrix3::kIdentity, Matrix3(1., 0., 0.,
+        0., 1., 0., 0., 0., 1.));
+    EXPECT_EQ(Matrix3::kOnes, std::initializer_list<double>({1., 1., 1.,
+        1., 1., 1., 1., 1., 1.}));
+    EXPECT_EQ(Matrix3::kZero, Matrix3(0., 0., 0.,
+        0., 0., 0., 0., 0., 0.));
+
+    EXPECT_EQ(m2 - m3, Matrix3(0., 0., 0.,
+        0., 0., 0., 0., 0., 0.));
+    EXPECT_EQ(m2 + m3, Matrix3(2., 4., 6.,
+        8., 10., 12., 14., 16., 18.));
+    EXPECT_EQ(m2 * m3, std::initializer_list<double>({1., 4., 9.,
+        16., 25., 36., 49., 64., 81.}));
+    EXPECT_EQ(m2 / m3, std::initializer_list<double>({1., 1., 1.,
+        1., 1., 1., 1., 1., 1.}));
+    EXPECT_EQ(m2 * 2, Matrix3(2., 4., 6.,
+        8., 10., 12., 14., 16., 18.));
+    EXPECT_EQ(2 * m2, std::initializer_list<double>({2., 4., 6.,
+        8., 10., 12., 14., 16., 18.}));
+    EXPECT_EQ(m2 * Vector3(1., 0., 0.), Vector3(1., 4., 7.));
+    EXPECT_NEAR(m2.det(), 0., kTolerance);
+
+    EXPECT_EQ(m1 += m3, Matrix3(2., 4., 6.,
+        8., 10., 12., 14., 16., 18.));
+    EXPECT_EQ(m1, std::initializer_list<double>({2., 4., 6.,
+        8., 10., 12., 14., 16., 18.}));
+    EXPECT_EQ(m1 -= m3, Matrix3(1., 2., 3.,
+        4., 5., 6., 7., 8., 9.));
+    EXPECT_EQ(m1, std::initializer_list<double>({1., 2., 3.,
+        4., 5., 6., 7., 8., 9.}));
+    EXPECT_EQ(m1 *= m3, Matrix3(1.,  4.,  9.,
+        16., 25., 36., 49., 64., 81.));
+    EXPECT_EQ(m1, std::initializer_list<double>({1.,  4.,  9.,
+        16., 25., 36., 49., 64., 81.}));
+    EXPECT_EQ(m1 /= m3, Matrix3(1., 2., 3.,
+        4., 5., 6., 7., 8., 9.));
+    EXPECT_EQ(m1, std::initializer_list<double>({1., 2., 3.,
+        4., 5., 6., 7., 8., 9.}));
+    EXPECT_EQ(m1 *= 2, Matrix3(2., 4., 6.,
+        8., 10., 12., 14., 16., 18.));
+    EXPECT_EQ(m1, std::initializer_list<double>({2., 4., 6.,
+        8., 10., 12., 14., 16., 18.}));
+    EXPECT_EQ(m1 /= 2, Matrix3(1., 2., 3.,
+        4., 5., 6., 7., 8., 9.));
+    EXPECT_EQ(m1, std::initializer_list<double>({1., 2., 3.,
+        4., 5., 6., 7., 8., 9.}));
+
+    m1 = Matrix3(1.,  4.,  9., 16., 25., 36., 49., 64., 81.);
+    EXPECT_TRUE(m1 == m2 * m3);
+    EXPECT_TRUE(m1 != m2);
+    EXPECT_TRUE(m1 == std::initializer_list<double>({1.,  4.,  9.,
+        16., 25., 36., 49., 64., 81.}));
+    EXPECT_TRUE(m1 != std::initializer_list<double>({1., 2., 3.,
+        4., 5., 6., 7., 8., 9.}));
+
+    std::stringstream ss;
+    ss << m3;
+    EXPECT_EQ(ss.str(), "[[1, 2, 3], [4, 5, 6], [7, 8, 9]]");
+
+    const std::vector<Vector3> kExpectedRows{Vector3(1., 2., 3.),
+        Vector3(4., 5., 6.), Vector3(7., 8., 9.)};
+    const std::vector<Vector3> kExpectedCols{Vector3(1., 4., 7.),
+        Vector3(2., 5., 8.), Vector3(3., 6., 9.)};
+
+    for (const Vector3& r : kExpectedRows) {
+      int found{0};
+      for (int i = 0; i < 3; ++i) {
+        if (r == m2.row(i)) {
```

Comment:
Isn't it enough to test that `m2.row(0) == kExpectedRows[0]` and so and so? Here we are testing that we find the vectors we expected, but not necessarily in the order we think they should be.

Reply:
Agree